src/analyses/loopTermination.ml

```diff
@@ -0,0 +1,130 @@
+
+(** Checks whether a variable can be bounded *)
+let check_bounded ctx varinfo =
+  let open IntDomain.IntDomTuple in
+  let exp = Lval (Var varinfo, NoOffset) in
+  match ctx.ask (EvalInt exp) with
+  | `Top -> false
+  | `Lifted v -> not (is_top_of (ikind v) v)
+  | `Bot -> failwith "Loop counter variable is Bot."
+
+(** We want to record termination information of loops and use the loop
+ * statements for that. We use this lifting because we need to have a
+ * lattice. *)
+module Statements = Lattice.Flat (CilType.Stmt) (Printable.DefaultNames)
+
+(** The termination analysis considering loops and gotos *)
+module Spec : Analyses.MCPSpec =
+struct
+
+  (** Provides some default implementations *)
+  include Analyses.IdentitySpec
+
+  let name () = "termination"
+
+  module D = Lattice.Unit
+  module C = D
+  module V =
+  struct
+    include Printable.Unit
+    let is_write_only _ = true
+  end
+  module G = MapDomain.MapBot (Statements) (BoolDomain.MustBool)
+
+  let startstate _ = ()
+  let exitstate = startstate
+
+  let finalize () =
+    (* Warning for detected possible non-termination *)
+    (* Upjumping gotos *)
+    if not (no_upjumping_gotos ()) then (
+      List.iter
+        (fun x ->
+           let msgs =
+             [(Pretty.dprintf
+                 "The program might not terminate! (Upjumping Goto)",
+               Some (M.Location.CilLocation x)
+              );] in
+           M.msg_group Warning ~category:NonTerminating "Possibly non terminating loops" msgs)
+        (!upjumping_gotos)
+    );
+    (* Multithreaded *)
+    if not (!single_thread) then (
+      M.warn ~category:NonTerminating "The program might not terminate! (Multithreaded)\n"
+    )
+
+  let special ctx (lval : lval option) (f : varinfo) (arglist : exp list) =
+    if !AnalysisState.postsolving then
+      match f.vname, arglist with
+        "__goblint_bounded", [Lval (Var x, NoOffset)] ->
+        (try
+           let loop_statement = VarToStmt.find x !loop_counters in
+           let is_bounded = check_bounded ctx x in
+           ctx.sideg () (G.add (`Lifted loop_statement) is_bounded (ctx.global ()));
+           (* In case the loop is not bounded, a warning is created. *)
+           if not (is_bounded) then (
+             let msgs =
+               [(Pretty.dprintf
+                   "The program might not terminate! (Loop analysis)",
+                 Some (M.Location.CilLocation (Cilfacade.get_stmtLoc loop_statement))
+                );] in
+             M.msg_group Warning ~category:NonTerminating "Possibly non terminating loops" msgs);
+           ()
+         with Not_found ->
+           (* This should not happen as long as __goblint_bounded is only used
+            * for this analysis. *)
+           ())
+      | _ -> ()
+    else ()
+
+  (** Checks whether a new thread was spawned some time. We want to discard
+   * any knowledge about termination then (see query function) *)
+  let must_be_single_threaded_since_start ctx =
```

Comment:
Does the `_since_start` suffix in the name still fit to the implementation? Maybe one can just remove that suffix in the name?

Reply:
We can. I think it is still fitting as the information should not be confused with single-threadedness at a certain point of execution.